Include Highlight in the bundle

Tatsh opened the pull request:
This will fix #72.

I don't know if you want to keep your own forks of Lua and highlight because their build systems are not so fun. Especially Lua because it wants users to modify `makefile` as they need to. They do not want to use something more configurable like autoconf or CMake (or Xcode).

![image](https://user-images.githubusercontent.com/724848/101279226-d007fc00-378e-11eb-9ef8-838982f64495.png)

This should work like so:

```sh
git clone --recurse-submodules -b bigsur 'git@github.com:Tatsh/QLColorCode'
cd QLColorCode
xcodebuild -target Package
qlmanage -r cache
```

Then preview a supported file in Finder.

ralphrmartin commented:
Can't try it out as I get 
git@gitlab.com: Permission denied (publickey,keyboard-interactive).
fatal: Could not read from remote repository.
...
fatal: clone of 'git@gitlab.com:saalen/highlight.git' into submodule path ...failed

Perhaps it means I need a gitlab account?

Tatsh commented:
I switched the module URIs to be https which allows anonymous cloning. Should fix the issue.

ralphrmartin commented:
Unfortunately the xcodebuild step failed:

```
 $ xcodebuild -target Package
Command line invocation:
    /Applications/Xcode.app/Contents/Developer/usr/bin/xcodebuild -target Package
...
In file included from ./core/texgenerator.cpp:32:
In file included from ./include/texgenerator.h:33:
In file included from ./include/codegenerator.h:37:
./include/syntaxreader.h:39:10: fatal error: 'boost/xpressive/xpressive_dynamic.hpp' file not found
#include <boost/xpressive/xpressive_dynamic.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I already have boost 1.74.0 installed via homebrew.

Tatsh commented:
Can you tell me the path to Boost include? I assume it's under /usr/local/Cellar or something but I don't know the full path.

I use MacPorts and `/opt/local/include` is in my patch. This should be the directory where `boost/unordered_map.hpp` should exist (example). That is, subtracting `/boost*` from `/opt/local/include/boost/unordered_map.hpp` in my case which results in `/opt/local/include`.

You can try modifiying `hl/highlight/src/makefile` and editing line 78 to add `-I ${other include path}` after the other `-I` arguments.

ralphrmartin commented:
That particular include file is at 

/usr/local/Cellar/boost/1.74.0/include/boost/unordered_map.h

ralphrmartin commented:
OK, edited the makefile to replace /opt/local... with /usr/local... and QLColorCode built successfully, and seems to run OK.

Tatsh commented:
I can make the patch script add the Boost path dynamically but maybe it's better to include Boost as a submodule.

Tatsh commented:
> OK, edited the makefile to replace /opt/local... with /usr/local... and QLColorCode built successfully, and seems to run OK.

Could you post a screenshot similar to mine? With About Mac open.

ralphrmartin commented:
![Screenshot 2020-12-06 at 19 38 35](https://user-images.githubusercontent.com/5293194/101290529-cfa54a80-37fa-11eb-8887-74e1f8ca762c.jpg)

Tatsh commented:
I gave up on trying to add Boost as a submodule. I tried to only add the submodules necessary but this has become a crazy headache. If I add the main project it will add a huge number of submodules and then the set up for making the highlight source see these headers takes forever.

I added the `/usr/local/include` path to the patch for Highlight and added a note in the readme.

Please try the commands in the first post again and see if it works out of the box.

ralphrmartin commented:
Yes, builds without modification now.

I did note this, I don't know if it's harmless:

```
    /bin/sh -c /Users/ralph/junk/QLColorCode/build/QLColorCode.build/Release/Package.build/Script-0E0E34E80D0097020086C1E6.sh
+/Users/ralph/junk/QLColorCode/build/QLColorCode.build/Release/Package.build/Script-0E0E34E80D0097020086C1E6.sh:3> [ '!=' Release ']'
/Users/ralph/junk/QLColorCode/build/QLColorCode.build/Release/Package.build/Script-0E0E34E80D0097020086C1E6.sh:3: parse error: condition expected: !=
```

Tatsh commented:
Harmless for this but it needs to be looked at.

dnicolson commented:
I had to run `brew install boost` but otherwise it works well.

Tatsh commented:
> I had to run `brew install boost` but otherwise it works well.

Yes this is required for building but not runtime. The readme has been updated to state this.

ElfSundae commented:
Great job :+1:  build succeeded after `brew install boost`, on macOS 11.0.1

BrianGilbert commented:
Also worked for me after installing boost via homebrew 👍

ZhouyangFANG commented:
Thanks for the fix and hope this change can be released in Homebrew later so I can update with ease. Thanks a lot.

anthonygelibert commented:
THANKS A LOT TO EVERYBODY!!!
